File: src/main/java/pw/vodes/rimuru/file/AutoMod.java
package pw.vodes.rimuru.file;

import java.time.Duration;
import java.util.ArrayList;
import java.util.List;
import java.util.regex.Pattern;

import org.apache.commons.lang3.StringUtils;
import org.javacord.api.entity.channel.TextChannel;
import org.javacord.api.entity.message.Message;
import org.javacord.api.entity.message.MessageAuthor;
import org.javacord.api.entity.message.embed.EmbedBuilder;
import org.javacord.api.event.message.MessageCreateEvent;
import org.javacord.api.event.message.MessageEditEvent;
import org.javacord.api.listener.message.MessageCreateListener;
import org.javacord.api.listener.message.MessageEditListener;

import com.fasterxml.jackson.core.JsonProcessingException;

import pw.vodes.rimuru.Main;
import pw.vodes.rimuru.file.sub.AutoModFile;
import pw.vodes.rimuru.file.sub.AutoModFile.Punishment;

public class AutoMod {
	
	private static List<AutoModFile> automods = new ArrayList<AutoModFile>();
	private static ArrayList<TextChannel> excludedChannels = new ArrayList<TextChannel>();
	
	@SuppressWarnings("unchecked")
	public static void load() {
		if(Main.getFiles().exists("automod.json")) {
			try {
				automods = (ArrayList<AutoModFile>)Main.getMapper().readValue(Main.getFiles().read("automod.json"), Main.getMapper().getTypeFactory().constructCollectionType(ArrayList.class, AutoModFile.class));
			} catch (JsonProcessingException e) {
				e.printStackTrace();
			}
		}
		
		for(var channel : Main.getServer().getTextChannels()) {
			for(var cat : Main.getConfig().getAutomodExcludedCategories()) {
				if(channel.asCategorizable().get().getCategory().isPresent()) {
					if(channel.asCategorizable().get().getCategory().get().getName().equalsIgnoreCase(cat)) {
						excludedChannels.add(channel);
					}
				}
			}
		}
	}
	
	private static void save() {
		try {
			Main.getFiles().write("automod.json", Main.getMapper().writerWithDefaultPrettyPrinter().writeValueAsString(automods));
		} catch (JsonProcessingException e) {
			e.printStackTrace();
		}
	}
	
	public static void add(AutoModFile am) {
		automods.add(am);
		save();
	}
	
	public static List<AutoModFile> getAutomods() {
		return automods;
	}
	
	public static void remove(AutoModFile am) {
		automods.remove(am);
		save();
	}
	
	public static MessageCreateListener getAutomodCreateListener() {
		return new MessageCreateListener() {
			@Override
			public void onMessageCreate(MessageCreateEvent event) {
				checkMessage(event.getMessage(), event.getMessageAuthor(), event.getMessageContent());
			}
		};
	}
	
	public static MessageEditListener getAutomodEditListener() {
		return new MessageEditListener() {

			@Override
			public void onMessageEdit(MessageEditEvent event) {
				var msg = event.getMessage().get();
				checkMessage(msg, msg.getAuthor(), msg.getContent());				
			}
		};
	}
	
	private static void checkMessage(Message message, MessageAuthor author, String content) {
		var isAdminOrMod = author.isServerAdmin();
		for(var role : Main.getConfig().getModRoles()) {
			if(role.hasUser(author.asUser().get()))
				isAdminOrMod = true;
		}
		
		if(isAdminOrMod)
			return;
		
		for(var chan : excludedChannels) {
			if(message.getChannel().getId() == chan.getId()) {
				return;
			}
		}
		
		for(var am : AutoMod.getAutomods()) {
			try {
				var pattern = Pattern.compile(am.filteredSequence, Pattern.CASE_INSENSITIVE);
				if(pattern.matcher(content).find() || StringUtils.containsIgnoreCase(content, am.filteredSequence)) {
					var embed = new EmbedBuilder().setTitle("Automod");
					embed.setAuthor(author);
					embed.setUrl(message.getLink().toString());
					embed.setFooter("UserID: " + author.getIdAsString());
					if(am.punishment == null) {
						embed.setDescription("Message deleted:\n```" + content + "```");
						message.delete();
					} else {
						if(am.punishment == Punishment.KICK) {
							Main.getServer().kickUser(author.asUser().get(), "Automod");
							message.delete();
							embed.setDescription("Kicked & Message deleted:\n```" + content + "```");
						} else if(am.punishment == Punishment.BAN) {
							Main.getServer().banUser(author.getId(), 0, "Automod");
							message.delete();
							embed.setDescription("Banned & Message deleted:\n```" + content + "```");
						} else {
							Main.getServer().timeoutUser(author.asUser().get(), Duration.ofMinutes(Long.parseLong(am.punishmentVal)), "Automod");
							message.delete();
							embed.setDescription("Timeouted & Message deleted:\n```" + content + "```");
						}
					}
					Main.getConfig().getStaffActionChannel().sendMessage(embed);
					break;
				}
			} catch (Exception e) {}
		}
	}

}

File: src/main/kotlin/pw/vodes/rimurukt/file/AutoMod.kt
package pw.vodes.rimurukt.file

import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import org.javacord.api.entity.channel.ServerTextChannel
import org.javacord.api.entity.message.Message
import org.javacord.api.entity.message.MessageAuthor
import org.javacord.api.entity.message.embed.EmbedBuilder
import org.javacord.api.listener.message.MessageCreateListener
import org.javacord.api.listener.message.MessageEditListener
import pw.vodes.rimurukt.Main
import pw.vodes.rimurukt.json
import java.io.File
import java.time.Duration
import kotlin.jvm.optionals.getOrNull

@Serializable
data class AutoModEntry(var filteredSequence: String, var punishment: Punishment?, var punishmentVal: String)

enum class Punishment {
    TIMEOUT, KICK, BAN;

    fun get(name: String) = entries.find { name.equals(it.toString(), true) }
}

object AutoMod {
    private val file = File(Main.appDir, "automod.json")
    private val excludedChannels = mutableListOf<ServerTextChannel>()
    var automods = mutableListOf<AutoModEntry>()

    fun load() {
        if (file.exists()) {
            automods = json.decodeFromString(file.readText())
        }

        Main.server.textChannels.forEach { channel ->
            if (!channel.category.isPresent)
                return@forEach

            val name = channel.category.get().name
            if (Main.config.automodExcludedCategories.find { it.equals(name, true) } != null)
                excludedChannels.add(channel)
        }
    }

    fun save() {
        file.writeText(json.encodeToString(automods))
    }

    fun automodCreateListener(): MessageCreateListener {
        return MessageCreateListener { event -> checkMessage(event.message, event.messageAuthor, event.messageContent) }
    }

    fun automodEditListener(): MessageEditListener {
        return MessageEditListener { event -> checkMessage(event.message, event.messageAuthor, event.messageContent) }
    }

    private fun checkMessage(message: Message, author: MessageAuthor, content: String) {
        var isStaff = author.isServerAdmin
        val user = author.asUser().getOrNull() ?: return
        Main.config.modRoles().forEach {
            if (isStaff)
                return@forEach
            isStaff = it.hasUser(user)
        }
        if (isStaff || excludedChannels.find { it.id == message.channel.id } != null)
            return

        for (automod in automods) {
            val pattern = Regex(automod.filteredSequence, RegexOption.IGNORE_CASE)
            if (!pattern.matches(content) && !content.contains(automod.filteredSequence, true))
                continue

            val embed = EmbedBuilder().setTitle("Automod").setAuthor(author).setUrl(message.link.toString()).setFooter("UserID: ${author.id}")
            if (automod.punishment == null) {
                embed.setDescription("Message deleted:\n```$content```")
                message.delete()
            } else {
                when (automod.punishment!!) {
                    Punishment.KICK -> {
                        Main.server.kickUser(user, "Automod")
                        message.delete()
                        embed.setDescription("Kicked & Message deleted:\n```$content```")
                    }

                    Punishment.BAN -> {
                        Main.server.banUser(author.id, Duration.ZERO, "Automod")
                        message.delete()
                        embed.setDescription("Banned & Message deleted:\n```$content```")
                    }

                    Punishment.TIMEOUT -> {
                        Main.server.timeoutUser(user, Duration.ofMinutes(automod.punishmentVal.toLong()), "Automod")
                        message.delete()
                        embed.setDescription("Timeouted & Message deleted:\n```$content```")
                    }
                }
            }
            Main.config.userLogChannel()!!.sendMessage(embed)
            break
        }
    }
}

